lib/Index/Index.cpp

```diff
@@ -193,6 +193,17 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
     return false;
   }
 
+  bool getModuleUSR(ModuleEntity Mod, StringRef &USR) {
+    SmallString<128> storage;
+    {
+      llvm::raw_svector_ostream OS(storage);
+      if (ide::printModuleUSR(Mod, OS))
+        return true;
+      USR = stringStorage.copyString(OS.str());
+    }
+    return false;
+  }
```

Comment:
Things to change for `getModuleUSR()`:

* Change it to `getModuleNameAndUSR(ModuleEntity Mod, StringRef &name, StringRef &USR)`. On the clang side we record clang module references with their full name ("ClangModule.SubModule"), and to match this on the swift side you'd need to use `ModuleEntity::getFullName()` which constructs a new string that you'd also copy in `stringStorage`.
* This function, as it is now, creates a new USR string (and puts it in `stringStorage`) every time you call it. Note that the other functions, `getNameAndUSR()` and `getPseudoAccessorNameAndUSR()` associate the strings with the pointers, so if you call the functions with the same pointer, you get the strings directly without allocating new strings each time.
To do the same thing here I'd suggest:
    * Change `nameAndUSRCache` to have a key of `void *` so you can reuse it for module pointers as well.
    * In `ModuleEntity` add `void *getOpaqueValue() const` and have that just return 'Mod.getOpaqueValue()`
    * Follow what `getNameAndUSR()` is doing but for the key use `ModuleEntity::getOpaqueValue()`.